pkg/agent/controller/serviceexternalip/controller_test.go

```diff
@@ -0,0 +1,636 @@
+		clientset:                   clientset,
+		informerFactory:             informerFactory,
+		mockIPAssigner:              mockIPAssigner,
+		fakeMemberlistCluster:       memberlistCluster,
+	}
+}
+
+func makeService(name, namespace string, serviceType corev1.ServiceType,
+	trafficPolicy corev1.ServiceExternalTrafficPolicyType, ipPool, externalIP string) *corev1.Service {
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Type:                  serviceType,
+			ExternalTrafficPolicy: trafficPolicy,
+		},
+	}
+	if externalIP != "" {
+		service.Status.LoadBalancer.Ingress = []corev1.LoadBalancerIngress{
+			{IP: externalIP},
+		}
+	}
+	if ipPool != "" {
+		service.Annotations = map[string]string{
+			types.ExternalIPPoolAnnotationKey: ipPool,
+		}
+	}
+	return service
+}
+
+func makeEndpoints(name, namespace string, addresses, notReadyAddresses map[string]string) *corev1.Endpoints {
+	var addr, notReadyAddr []corev1.EndpointAddress
+	for k, v := range addresses {
+		ip := k
+		addr = append(addr, corev1.EndpointAddress{
+			IP:       ip,
+			NodeName: stringPtr(v),
+		})
+	}
+	for k, v := range notReadyAddresses {
+		ip := k
+		notReadyAddr = append(notReadyAddr, corev1.EndpointAddress{
+			IP:       ip,
+			NodeName: stringPtr(v),
+		})
+	}
+	service := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+
+		Subsets: []corev1.EndpointSubset{
+			{
+				Addresses:         addr,
+				NotReadyAddresses: notReadyAddr,
+			},
+		},
+	}
+	return service
+}
+
+func TestCreateService(t *testing.T) {
+	tests := []struct {
+		name                     string
+		existingEndpoints        []*corev1.Endpoints
+		serviceToCreate          *corev1.Service
+		healthyNodes             []string
+		expectedCalls            func(mockIPAssigner *ipassignertest.MockIPAssigner)
+		expectedExternalIPStates map[apimachinerytypes.NamespacedName]externalIPState
+		expectError              bool
+	}{
+		{
+			name:              "new Service created and local Node selected",
+			existingEndpoints: nil,
+			serviceToCreate:   servicePolicyCluster,
+			healthyNodes:      []string{fakeNode1, fakeNode2},
+			expectedCalls: func(mockIPAssigner *ipassignertest.MockIPAssigner) {
+				mockIPAssigner.EXPECT().AssignIP(fakeServiceExternalIP1)
+			},
+			expectedExternalIPStates: map[apimachinerytypes.NamespacedName]externalIPState{
+				keyFor(servicePolicyCluster): {
+					ip:           fakeServiceExternalIP1,
+					assignedNode: fakeNode1,
+				},
+			},
+			expectError: false,
+		},
+
+		{
+			name:              "new Service created and local Node not selected",
+			existingEndpoints: nil,
+			serviceToCreate:   servicePolicyCluster,
+			healthyNodes:      []string{fakeNode2, fakeNode1},
+			expectedCalls: func(mockIPAssigner *ipassignertest.MockIPAssigner) {
+				mockIPAssigner.EXPECT().UnassignIP(fakeServiceExternalIP1)
```

Comment:
Why it gets different result when all inputs are same and only healthyNodes change order?

Reply:
The order of `healthyNodes` also determins the hash order in the fake consistant hash in fake memberlist. I changed this to use a fake hash function to make the select Node predictable.

Reply:
I am not sure whether the `Members()` will get consistent results or not (I think the answer is no). But as we stored the hashed replaces in constitanthash, it should not be a problem. Initially, the testing code here uses the order of `healthy nodes` to determine the order of hashed values, thus making it obscure. I have added a fake hashing function here to make it more clear.